lib/capistrano/configuration/scm_resolver.rb

```diff
@@ -0,0 +1,149 @@
+    # won't be necessary once backwards compatibility is removed in a future
+    # version.
+    #
+    # TODO: Remove this class entirely in Capistrano 4.0.
+    #
+    class SCMResolver
+      DEFAULT_GIT = :"default-git"
+
+      include Capistrano::DSL
+
+      def resolve
+        return if scm_name.nil?
+        set(:scm, :git) if using_default_scm?
+
+        print_deprecation_warnings_if_applicable
+        return if scm_plugin_loaded?
+
+
+        if built_in_scm_name?
+          load_built_in_scm
+        else
+          # Compatibility with existing 3.x third-party SCMs
+          register_legacy_scm_hooks
+          load_legacy_scm_by_name
+        end
+      end
+
+      private
+
+      def using_default_scm?
+        return @using_default_scm if defined? @using_default_scm
+        @using_default_scm = (fetch(:scm) == DEFAULT_GIT)
+      end
+
+      # This is somewhat of a hack, because there is no guarantee that a third-
+      # party SCM will necessarily implement the typical SCM tasks. But it works
+      # well enough for the built-in SCMs.
+      def scm_plugin_loaded?
+        Rake::Task.tasks.any? { |t| t.name =~ /^[^:]+:create_release/ }
+      end
+
+      def scm_name
+        fetch(:scm)
+      end
+
+      def load_built_in_scm
+        require "capistrano/scm/#{scm_name}"
+        scm_class = Object.const_get(built_in_scm_plugin_class_name)
+        # We use :load_immediately because we are initializing the SCM plugin
+        # late in the load process and therefore can't use the standard
+        # load:defaults technique.
+        scm_class.new(:load_immediately => true)
+      end
+
+      def load_legacy_scm_by_name
+        load("capistrano/#{scm_name}.rb")
+      end
+
+      def third_party_scm_name?
+        !built_in_scm_name?
+      end
+
+      def built_in_scm_name?
+        %w(git hg svn).include?(scm_name.to_s.downcase)
+      end
+
+      def built_in_scm_plugin_class_name
+        "Capistrano::SCM::#{scm_name.to_s.capitalize}"
+      end
+
+      def register_legacy_scm_hooks
+        if Rake::Task.task_defined?("deploy:new_release_path")
+          after "deploy:new_release_path", "#{scm_name}:create_release"
+        end
+
+        if Rake::Task.task_defined?("deploy:check")
+          before "deploy:check", "#{scm_name}:check"
+        end
+
+        if Rake::Task.task_defined?("deploy:set_current_revision")
+          before "deploy:set_current_revision",
+                 "#{scm_name}:set_current_revision"
+        end
+      end
+
+      def print_deprecation_warnings_if_applicable
+        if using_default_scm?
+          warn_add_git_to_capfile unless scm_plugin_loaded?
+        elsif built_in_scm_name?
+          warn_set_scm_is_deprecated
+        elsif third_party_scm_name?
+          warn_third_party_scm_must_be_upgraded
+        end
+      end
+
+      def warn_set_scm_is_deprecated
+        $stderr.puts(<<-MESSAGE)
+WARNING: `set :scm, #{scm_name.inspect}` is deprecated.
```

Comment:
Would this be better as a NOTICE rather than a WARNING? I think most users would read this and think something broke...